src/main/scala/TypeCheck.scala

```diff
@@ -175,6 +176,15 @@ object TypeChecker {
     case EInt(v, _) => TStaticInt(v) -> env
     case EBool(_) => TBool() -> env
     case ERecLiteral(_) => throw RecLiteralNotInBinder(expr.pos)
+    case ECast(e, castType) => {
+      val (typ, nEnv) = checkE(e)
+      if (safeCast(typ, castType) == false) {
+        scribe.warn {
+          (s"Casting expression of type $typ to $castType which might not be safe", expr)
```

Comment:
One tiny suggestion here: it seems like the rules for emitting a warning are trying to avoid classic sources of precision loss. So a more specific phrasing could say, "Casting X to Y may lose precision."

Reply:
Addressed in 7f375d0